### PR TITLE
add public agent role

### PR DESCRIPTION
**Pull request by darkonie**
- add public agent role to installer.
- refactor installer code.
- update integration / unit tests.
- make agent_list and agent_public_list not required to pass validation.

### CODE REVIEW

**Comment by darkonie** [comment]
@cmaloney this is ready for code review

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py [hunk 1]
```diff
@@ -16,6 +16,26 @@ class ExecuteException(Exception):
     """Raised when execution fails"""
 
 
+def add_nodes(nodes, tag):
```
Comment (cmaloney) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: This is only used / called once. Why not make it a local or inline function?

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py [hunk 2]
```diff
@@ -65,7 +78,7 @@ def run_preflight(config, pf_script_path='/genconf/serve/dcos_install.sh', block
     chains.append(cleanup_chain)
     master_agent_count = {
         'total_masters': len(config['master_list']),
-        'total_agents': len(config['agent_list'])
+        'total_agents': len(config['agent_list']) + len(config['public_agent_list'])
```
Comment (cmaloney) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: This code comes up twice. Why not have a single function which you call to go from config -> nodes_count_by_type?

This could also potentially be more cleanly written using a python groupby (https://docs.python.org/3/library/itertools.html#itertools.groupby). But that is probably out of scope for this PR.

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py [hunk 3]
```diff
@@ -151,10 +164,10 @@ def _remove_host(state_file, host):
...
 
 @asyncio.coroutine
-def install_dcos(config, block=False, state_json_dir=None, hosts=[], async_delegate=None, try_remove_stale_dcos=False,
-                 roles=None, **kwargs):
-    if roles is None:
-        roles = ['master', 'agent']
+def install_dcos(config, block=False, state_json_dir=None, hosts=None, async_delegate=None, try_remove_stale_dcos=False,
+                 **kwargs):
+    if hosts is None:
+        hosts = []
```
Comment (cmaloney) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: Why did the default for hosts change to None? Shouldn't there just be an assertion that hosts is an instance of a list?

File: packages/dcos-installer-ui/buildinfo.json [hunk 4]
```diff
@@ -3,7 +3,7 @@
   "single_source": {
     "kind": "git",
     "git": "https://github.com/dcos/dcos-installer-ui.git",
-    "ref": "ba16865174fdf51a3f487112f5c390d500c155c1",
+    "ref": "b2b5ace30f33dbe3d983e6db91c10ecb0974b334",
```
Comment (cmaloney) [inline packages/dcos-installer-ui/buildinfo.json]: This should be updated / changed independently / not in this PR

File: test_util/run-all [hunk 5]
```diff
@@ -113,7 +114,7 @@ RET=$?
 mkdir -p $WORKDIR/test/vagrant-logs
 if [ $RET -ne 0 ]; then
   echo "TEST FAILED"
-  for i in boot m1 m2 m3 a1 a2; do
+  for i in boot m1 m2 m3 a1 a2 p2; do
```
Comment (cmaloney) [inline test_util/run-all]: why not p1?

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py [hunk 6]
```diff
@@ -164,22 +177,21 @@ def install_dcos(config, block=False, state_json_dir=None, hosts=[], async_deleg
         },
         'agent': {
             'tags': {'role': 'agent', 'dcos_install_param': 'slave'},
-            'hosts': hosts if hosts else config['agent_list']
+            'hosts': hosts if hosts else config.get('agent_list', [])
+        },
+        'public_agent': {
+            'tags': {'role': 'public_agent', 'dcos_install_param': 'slave_public'},
+            'hosts': hosts if hosts else config.get('public_agent_list', [])
```
Comment (cmaloney) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: This seems to say that if "hosts" is supplied we should always make them public_agents as well as private_agents

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py (hunk 1 above)
Comment (darkonie) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: done

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py (hunk 2 above)
Comment (darkonie) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: done

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py (hunk 3 above)
Comment (darkonie) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: basically it is a bad practice to have mutable default parameter, this is a workaround to make sure if `hosts` wasn't passed, we should always default to a new list.

File: test_util/run-all (hunk 5 above)
Comment (darkonie) [inline test_util/run-all]: done

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py (hunk 6 above)
Comment (darkonie) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: done

**Comment by darkonie** [comment]
@cmaloney this is ready for review. See a second commit. Once it is ready to ship i'll squash all commits.

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py [hunk 7]
```diff
@@ -151,35 +163,37 @@ def _remove_host(state_file, host):
 
 
 @asyncio.coroutine
-def install_dcos(config, block=False, state_json_dir=None, hosts=[], async_delegate=None, try_remove_stale_dcos=False,
-                 roles=None, **kwargs):
-    if roles is None:
-        roles = ['master', 'agent']
+def install_dcos(config, block=False, state_json_dir=None, hosts=None, async_delegate=None, try_remove_stale_dcos=False,
+                 **kwargs):
+    if hosts is None:
```
Comment (cmaloney) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: Can we add an assert that everything in the hosts list is an instance of Nodes if hosts is given? (This code will make really odd error messages if that isn't the case)

File: ssh/validate.py [hunk 8]
```diff
@@ -27,32 +27,35 @@ def validate_ssh_port(ssh_port):
...
+
+    if public_agent_list:
+        compare_lists(master_list, public_agent_list)
 
 
 def check_duplicates(arg_list):
     assert isinstance(arg_list, list), 'only lists can be verified for duplicates'
     dups = list(filter(lambda x: arg_list.count(x) > 1, arg_list))
-    assert len(dups) == 0, 'List cannot contain duplicates: {}'.format(', '.join(set(dups)))
+    assert not dups, 'List cannot contain duplicates: {}'.format(', '.join(set(dups)))
```
Comment (cmaloney) [inline ssh/validate.py]: nit: This (and a number of these condition checks) is an unrelated change

File: ssh/validate.py [hunk 9]
```diff
@@ -63,30 +66,55 @@ def check_ipv4_addrs(ips):
...
-        'master_list': lambda master_list: validate_master_agent_lists(master_list, config.get('agent_list'))
-    }
+    if public_agent_list:
+        validate_hosts_list(public_agent_list, 'public_agent_list')
+
+    if agent_list and public_agent_list:
+        compare_lists(agent_list, public_agent_list)
 
+
+def run_validate_config(value_func_map, config, required=True):
```
Comment (cmaloney) [inline ssh/validate.py]: `dev validate_config_chunk(key_validate_fn_map, config, keys_required)`
 - Make the function name closer to what the function is doing
 - Always require setting required. We don't call this from many places, defaults should be avoided

File: ssh/validate.py [hunk 10]
```diff
@@ -63,30 +66,55 @@ def check_ipv4_addrs(ips):
...
     errors = {}
-    for ssh_key, validate_func in ssh_keys_checks_map.items():
+    for ssh_key, validate_func in value_func_map.items():
         input_value = config.get(ssh_key)
```
Comment (cmaloney) [inline ssh/validate.py]:
```python
if ssh_key not in config:
  if required:
    errors[ssh_key] = ...
try:
  validate_func(config[ssh_key])
except:
...
```

File: ssh/validate.py [hunk 11]
```diff
@@ -63,30 +66,55 @@ def check_ipv4_addrs(ips):
...
+def validate_config(config):
+    assert isinstance(config, dict)
+
+    ssh_keys_checks_map_required = {
+        'ssh_user': validate_ssh_user,
+        'ssh_port': validate_ssh_port,
+        'ssh_key_path': validate_ssh_key_path,
+        'master_list': lambda master_list: validate_master_agent_lists(
+            master_list,
+            config.get('agent_list'),
```
Comment (cmaloney) [inline ssh/validate.py]:
```
config.get('agent_list', [])
config.get('public_agent_list', [])
```

File: ssh/validate.py [hunk 12]
```diff
@@ -63,30 +66,55 @@ def check_ipv4_addrs(ips):
...
-
         try:
             validate_func(input_value)
         except AssertionError as err:
             errors[ssh_key] = str(err)
     return errors
+
+
+def validate_config(config):
+    assert isinstance(config, dict)
```
Comment (cmaloney) [inline ssh/validate.py]: Might make sense when loading config to do a 
```
config.setdefault('agent_list', [])
config.setdefault('public_agent_list', [])
```

That way all the code everywhere can just assume master_list, public_agent_list, and agent_ilst always exist.

File: test_util/integration_test.py [hunk 13]
```diff
@@ -1004,7 +1008,7 @@ def test_3dt_nodes(cluster):
         assert len(response) == 1, 'nodes response must have only one field: nodes'
         assert 'nodes' in response
         assert isinstance(response['nodes'], list)
-        assert len(response['nodes']) == len(cluster.masters + cluster.slaves), (
+        assert len(response['nodes']) == len(cluster.masters + cluster.slaves) + len(cluster.public_slaves), (
```
Comment (cmaloney) [inline test_util/integration_test.py]: len(cluster.all_slaves) + len(cluster.masters)?

File: test_util/run-all [hunk 14]
```diff
@@ -113,7 +114,7 @@ RET=$?
 mkdir -p $WORKDIR/test/vagrant-logs
 if [ $RET -ne 0 ]; then
   echo "TEST FAILED"
-  for i in boot m1 m2 m3 a1 a2; do
+  for i in boot m1 m2 m3 a1 a2 p1; do
```
Comment (cmaloney) [inline test_util/run-all]: note: Will need to update the CI log collector as well that outputs to the actual TeamCity log in chunks

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py (hunk 7 above)
Comment (darkonie) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: done

File: ssh/validate.py (hunk 12 above)
Comment (darkonie) [inline ssh/validate.py]: done

File: test_util/run-all (hunk 14 above)
Comment (mellenburg) [inline test_util/run-all]: @cmaloney I just went ahead and checked in the CI log collection: https://github.com/dcos/dcos/pull/77

**Comment by spahl** [comment]
This is good now.

:ship:

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py [hunk 15]
```diff
@@ -151,35 +163,38 @@ def _remove_host(state_file, host):
 
 
 @asyncio.coroutine
-def install_dcos(config, block=False, state_json_dir=None, hosts=[], async_delegate=None, try_remove_stale_dcos=False,
-                 roles=None, **kwargs):
-    if roles is None:
-        roles = ['master', 'agent']
+def install_dcos(config, block=False, state_json_dir=None, hosts=None, async_delegate=None, try_remove_stale_dcos=False,
```
Comment (cmaloney) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: When this is called via a retry in async_server.py, hosts is passed through as a raw list of hosts, not a list of Node objects, which means later the `targets=` sometimes results in targets being set to a list of Node instances, and sometimes a list of whatever was in the json state.

**Comment by cmaloney** [comment]
@spahl @darkonie NAK for landing now, retrying is broken.

File: ext/dcos-installer/dcos_installer/action_lib/__init__.py (hunk 15 above)
Comment (spahl) [inline ext/dcos-installer/dcos_installer/action_lib/__init__.py]: Nice catch. I'm gonna back it out of the current train.

**Comment by darkonie** [comment]
@spahl @cmaloney eventually it will be converted into Node https://github.com/dcos/dcos/blob/master/ssh/ssh_runner.py#L78

this may be very implicit and not documented very well, but the idea was to be able to pass a list of ip addresses or a list of Node objects. So this will work correctly. Let me know if you want to me to fix it or leave it as is.

**Comment by cmaloney** [comment]
That is a very complicated / confusing code path... Please simplify. We want code to be followable / easily understood when reading it.

Also note that final / late `Node()` conversion doesn't attach the `master` or `slave` tag to the hosts, which I'm suspect would mean https://github.com/dcos/dcos/pull/17/files#diff-6c2af57e1ff0e55764fc9f3f45ab4ce5R222 would go wrong / break.

**Comment by darkonie** [comment]
@cmaloney @spahl done, ready for review

**Comment by cmaloney** [comment]
This is looking good to me. :ship: when you're happy with it.

**Comment by spahl** [comment]
@darkonie you can rebase and merge
